napari/benchmarks/utils.py

```diff
@@ -1,6 +1,167 @@
+from functools import lru_cache
+from typing import Literal, Optional, Sequence, Tuple, Union, overload
+
+import numpy as np
+from skimage import morphology
+
+
 class Skiper:
     def __init__(self, func):
         self.func = func
 
     def __contains__(self, item):
         return self.func(item)
+
+
+def _generate_ball(radius: int, ndim: int) -> np.ndarray:
+    """
+    Generate a ball of given radius and dimension.
+
+    Parameters
+    ----------
+    radius : int
+        Radius of the ball.
+    ndim : int
+        Dimension of the ball.
+
+    Returns
+    -------
+    ball : ndarray of uint8
+        Binary array of the hyper ball.
+    """
+
+    if ndim == 2:
+        return morphology.disk(radius)
+    if ndim == 3:
+        return morphology.ball(radius)
+    shape = (2 * radius + 1,) * ndim
+    radius_sq = radius**2
+    coords = np.indices(shape) - radius
+    return (np.sum(coords**2, axis=0) <= radius_sq).astype(np.uint8)
+
+
+def _generate_density(radius: int, ndim: int) -> np.ndarray:
+    """
+    Generate gaussian density of given radius and dimension.
+    """
+    shape = (2 * radius + 1,) * ndim
+    coords = np.indices(shape) - radius
+    dist = np.sqrt(np.sum(coords**2 / ((radius / 4) ** 2), axis=0))
+    res = np.exp(-dist)
+    res[res < 0.02] = 0
+    return res
+
+
+def _add_structure_on_coordinates(
+    data, points, shape, structure, values, assign_operator
+):
+    """
+    helper function to update data with structure at given coordinates
+    """
```

Comment:
What is `structure`? What is `assign_operator`? Can you write a complete docstring here? Even though it's private, it will help others understand the code in the future.

Reply:
added